File: src/linkedList/ts/linkedList.ts
import { INode } from "./types"
import { EmptyLinkedListError, InvalidFieldError, NotFoundError } from "./utils"

export class Node<Type> implements INode<Type> {
    constructor(
        public readonly id: number,
        public value: Type,
        public next: INode<Type> | null
    ) {}
}

export class LinkedList<Type> {
    private nextId: number = 0
    private head: INode<Type> | null

    constructor() {
        this.head = null
    }

    public addFirst(value: Type): number {
        this.nextId += 1
        const node = new Node(this.nextId, value, null)

        if (this.head !== null) {
            this.head = node
        } else {
            node.next = this.head
            this.head = node
        }

        return node.id
    }

    public addLast(value: Type): number {
        this.nextId += 1
        const node = new Node(this.nextId, value, null)

        if (this.head === null) {
            this.head = node
        } else {
            let current = this.head
            while (current?.next !== null) {
                current = current?.next
            }
            current.next = node
        }

        return node.id
    }

    public addAfter(value: Type, targetId: number): number {
        if (typeof targetId !== "number") {
            throw new InvalidFieldError("TargetId")
        } else if (this.head === null) {
            throw new EmptyLinkedListError()
        }

        this.nextId += 1
        let current = this.head

        while (current.id !== targetId) {
            if (current.next === null) {
                throw new NotFoundError()
            }
            current = current.next
        }

        const node = new Node<Type>(this.nextId, value, current.next)
        current.next = node

        return node.id
    }

    public addBefore(value: Type, targetId: number): number {
        if (typeof targetId !== "number") {
            throw new InvalidFieldError("TargetId")
        } else if (this.head === null) {
            throw new EmptyLinkedListError()
        }

        this.nextId += 1
        let previous = null
        let current = this.head

        while (current.id !== targetId) {
            if (current.next === null) {
                throw new NotFoundError()
            }
            previous = current
            current = current.next
        }

        const node = new Node<Type>(this.nextId, value, current)
        previous ? (previous.next = node) : (this.head = node)

        return node.id
    }

    public remove(targetId: number): Type {
        if (typeof targetId !== "number") {
            throw new InvalidFieldError("TargetId")
        } else if (this.head === null) {
            throw new EmptyLinkedListError()
        }

        let previous = null
        let current = this.head

        while (current.id !== targetId) {
            if (current.next === null) {
                throw new NotFoundError()
            }
            previous = current
            current = current.next
        }

        if (previous) {
            previous.next = current.next
        } else {
            this.head = current.next
        }

        return current.value
    }

    public update(targetId: number, value: Type): Pick<INode<Type>, "id" | "value"> {
        const foundNode = this.search(targetId)
        foundNode.value = value
        return { id: foundNode.id, value: foundNode.value }
    }

    public reverse(): void {
        let current = this.head
        let previous = null

        while (current) {
            const next = current.next
            current.next = previous

            previous = current
            current = next
        }

        this.head = previous
    }

    public isEmpty(): boolean {
        return this.head === null
    }

    public find(targetId: number): Pick<INode<Type>, "id" | "value"> {
        const foundNode = this.search(targetId)
        return { id: foundNode.id, value: foundNode.value }
    }

    public print(): void {
        console.log(JSON.stringify(this.head, null, 2))
    }

    private search(targetId: number): INode<Type> {
        if (typeof targetId !== "number") {
            throw new InvalidFieldError("TargetId")
        } else if (this.head === null) {
            throw new EmptyLinkedListError()
        }

        let current = this.head
        while (current.id !== targetId) {
            if (current.next === null) {
                throw new NotFoundError()
            }
            current = current.next
        }

        return current
    }
}

File: src/linkedList/c/linkedList.c
#include <stdio.h>
#include <stdlib.h>
#include "linkedList.h"

struct LinkedList* createLinkedList(){
    struct LinkedList *list = malloc(sizeof(struct LinkedList));
    list->nextId = 0;
    list->head = NULL;

    return list;
}

int addFirst(int value, struct LinkedList *list){
    struct Node *node = malloc(sizeof(struct Node));
    list->nextId +=1;

    node->id = list->nextId;
    node->value = value;
    node->next = list->head;

    list->head = node;
    return node->id;
}

int addLast(int value, struct LinkedList *list){
    struct Node *node = malloc(sizeof(struct Node));
    struct Node *current = list->head;
    list->nextId+=1;

    node->id = list->nextId;
    node->value = value;
    node->next = NULL;

    if (current == NULL){
        list->head = node;
    } else {
        while (current->next != NULL){
            current = current->next;
        }
        current->next = node;
    }

    return node->id;
}

int addBefore(int targetId, int value, struct LinkedList *list){
    if (isEmpty(list)){
        printf("Linked list is empty\n");
        return -1;
    }

    struct Node *previous = NULL;
    struct Node *current = list->head;

    while (current->id != targetId){
        if (current->next == NULL){
            printf("Node not found\n");
            return -1;
        }
        previous = current;
        current = current->next;
    }

    struct Node *node = malloc(sizeof(struct Node));
    list->nextId +=1;
    node->id = list->nextId;
    node->value = value;
    node->next = current;

    if (previous == NULL){
        list->head = node;
    } else {
        previous->next = node;
    }

    return node->id;
}

int addAfter(int targetId, int value, struct LinkedList *list){
    struct Node *foundNode = searchNode(targetId, list);
    if  (!foundNode){
        return -1;
    }

    struct Node *node = malloc(sizeof(struct Node));
    list->nextId +=1;
    node->id = list->nextId;
    node->value = value;
    node->next = foundNode->next;
    foundNode->next = node;

    return node->id;
}

int removeNode(int targetId, struct LinkedList *list){
    if (isEmpty(list)){
        printf("Linked list is empty\n");
        return -1;
    }

    struct Node *previous = NULL;
    struct Node *current = list->head;

    while (current->id != targetId){
        if (current->next == NULL){
            printf("Node not found\n");
            return -1;
        }
        previous = current;
        current = current->next;
    }

    if (previous == NULL){
        list->head = current->next;
    } else{
        previous->next = current->next;
    }

    free(current);
    return 0;
}

int getNode(int targetId, struct LinkedList *list){
    struct Node *foundNode = searchNode(targetId, list);
    if(!foundNode){
        return -1;
    }

    return foundNode->value;
}

int updateNode(int targetId, int value, struct LinkedList *list){
    struct Node *foundNode = searchNode(targetId, list);
    if(!foundNode){
        return -1;
    }

    foundNode->value = value;
    return 0;
}

int reverse(struct LinkedList *list){
    if (isEmpty(list)){
        printf("Linked list is empty\n");
        return -1;
    }

    struct Node *listRef = list->head;
    struct Node *current;
    list->head = NULL;

    while (listRef != NULL) {
        struct Node *node = malloc(sizeof(struct Node));
        node->id = listRef->id;
        node->value = listRef->value;
        node->next = NULL;

        if (list->head == NULL){
            list->head = node;
        } else {
            node->next = list->head;
            list->head = node;
        }
        current = listRef;
        listRef = listRef->next;
        free(current);
    }

    return 0;
}

int isEmpty(struct LinkedList *list){
    return list->head == NULL ? 1 : 0;
}

int print(struct LinkedList *list){
    if (isEmpty(list)){
        printf("%p\n", NULL);
        return -1;
    }

    struct Node *current = list->head;

    while (current != NULL){
        printf("%p - {id: %d, value: %d next: %p}\n", current, current->id, current->value, current->next);
        current = current->next;
    }
    printf("\n");

    return 0;
}

static struct Node* searchNode(int targetId, struct LinkedList *list){
    if (isEmpty(list)){
        printf("Linked list is empty\n");
        return NULL;
    }

    struct Node *current = list->head;

    while (current->id != targetId){
        if (current->next == NULL){
            printf("Node not found\n");
            return NULL;
        }
        current = current->next;
    }

    return current;
}
